src/Exthost/Msg.re

```diff
@@ -240,10 +256,51 @@ module LanguageFeatures = {
       })
     | Unregister({handle: int});
 
+  let parseDocumentSelector = json => {
+    Json.Decode.(
+      json |> Json.Decode.decode_value(list(DocumentFilter.decode))
+    );
+  };
+
   let handle = (method, args: Yojson.Safe.t) => {
     switch (method, args) {
     | ("$unregister", `List([`Int(handle)])) =>
       Ok(Unregister({handle: handle}))
+
+    | ("$registerDefinitionSupport", `List([`Int(handle), selectorJson])) =>
+      selectorJson
+      |> parseDocumentSelector
+      |> Result.map(selector => {
+           RegisterDefinitionSupport({handle, selector})
+         })
+      |> Result.map_error(Json.Decode.string_of_error)
+    | ("$registerDeclarationSupport", `List([`Int(handle), selectorJson])) =>
+      selectorJson
+      |> parseDocumentSelector
+      |> Result.map(selector => {
+           RegisterDeclarationSupport({handle, selector})
+         })
+      |> Result.map_error(Json.Decode.string_of_error)
+    | (
+        "$registerImplementationSupport",
+        `List([`Int(handle), selectorJson]),
+      ) =>
+      selectorJson
+      |> parseDocumentSelector
+      |> Result.map(selector => {
+           RegisterImplementationSupport({handle, selector})
+         })
+      |> Result.map_error(Json.Decode.string_of_error)
+    | (
+        "$registerTypeDefinitionSupport",
+        `List([`Int(handle), selectorJson]),
+      ) =>
+      selectorJson
+      |> parseDocumentSelector
+      |> Result.map(selector => {
+           RegisterImplementationSupport({handle, selector})
+         })
+      |> Result.map_error(Json.Decode.string_of_error)
```

Comment:
If you're mapping both cases it's usually clearer, and sometimes also shorter, to just use a plain old `switch`

```suggestion
      switch (parseDocumentSelector(selectorJson)) {
      | Ok(selector) => RegisterImplementationSupport({handle, selector})
      | Error(error) => Json.Decode.string_of_error(error)
      }
```